ui/src/QrDataProvider.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { createContext, useContext, useReducer } from 'react';
+
+interface State {
+  qrCodes?: null | string[];
+  jws?: null | string;
+  setQrCodes: (arg: any) => any;
+  resetQrCodes: () => any;
+}
+
+let initialState: State = {
+  qrCodes: null,
+  jws: null,
+  setQrCodes: () => null,
+  resetQrCodes: () => null
+};
+
+const actions = {
+  SET_QR_CODES: 'SET_QR_CODES',
+  RESET_QR_CODES: 'RESET_QR_CODES',
+};
+
+const QrDataContext = createContext<State>(initialState);
+
+const getJws = (qrCodes: string[]) => {
+  return qrCodes.map((c) => {
+    const sliceIndex = c.lastIndexOf('/');
+  const rawPayload = c.slice(sliceIndex + 1);
+  const encodingChars = rawPayload.match(/\d\d/g);
+  return encodingChars?.map((charPair) => String.fromCharCode(+charPair + 45)).join('');
+  }).join('');
+};
+
+const reducer = (state: any, action: any) => {
+  console.log(localStorage);
+  switch (action.type) {
+    case actions.SET_QR_CODES: {
+      let newState: State = initialState;
+      localStorage.setItem('qrCodes', JSON.stringify(action.qrCodes));
+
+      if (action.qrCodes) {
+        newState.qrCodes = action.qrCodes;
+        newState.jws = '';
+        const jws = getJws(action.qrCodes);
+        newState.jws = jws;
+      } else newState.jws = null;
+      return {
+        ...state,
+        ...newState,
+      };
+    }
+    case actions.RESET_QR_CODES: {
+      localStorage.setItem('qrCodes', '');
+      return initialState;
+    }
+    default:
+      return state;
+  }
+};
+
+const QrDataProvider = ({ children }: any) => {
+  const [state, dispatch] = useReducer(
+    reducer,
+    reducer(initialState, {
+      type: actions.SET_QR_CODES,
+      qrCodes: JSON.parse(localStorage.getItem('qrCodes') || ''),
+    }),
+  );
+
+  const value = {
+    qrCodes: state.qrCodes,
+    jws: state.jws,
+    validationStatus: state.validationStatus,
+    matchingDemographicData: state.matchingDemographicData,
+    setQrCodes: (qrCodes: string) => {
+      dispatch({ type: actions.SET_QR_CODES, qrCodes });
+    },
+    resetQrCodes: () => {
+      dispatch({ type: actions.RESET_QR_CODES });
+    },
+  };
+  return <QrDataContext.Provider value={value}>{children}</QrDataContext.Provider>;
+};
+
+const useQrDataContext = () => useContext(QrDataContext);
+
+export { QrDataContext, QrDataProvider, useQrDataContext };
```

Comment:
Have you considered exposing just a a `useQrStore()` function here, returning just the fields in `value` above and making this object available as something like `scannedQrStore` in existing app-level context in https://github.com/jmandel/vaxx.link/blob/e4a98ae5394118d5b2618b12ebbb591c49a610f5/ui/src/App.tsx#L825 ?

This would help avoid the ceremony of a standalone context provider and avoid the need to wrap routes in `index.tsx`, and would help make this store look like and sit next to the existing app store.